### PR TITLE
SM Client

### PR DESCRIPTION
**Pull request by al1img**

### CODE REVIEW

**Review by MykolaSolyanko — APPROVED** [review]

File: src/clocksync/clocksync.hpp [hunk 1]
```diff
@@ -53,15 +63,29 @@ class ClockSyncItf {
...
      */
-    virtual aos::Error Sync(const aos::Time& time) = 0;
+    virtual Error Sync(const aos::Time& time) = 0;
+
+    /**
+     * Subscribes for clock sync notifications.
+     *
+     * @param subscriber clock sync notification subscriber.
+     */
+    virtual Error Subscribes(ClockSyncSubscriberItf& subscriber) = 0;
```
Comment (mlohvynenko) [inline src/clocksync/clocksync.hpp:81]:
```suggestion
    virtual Error Subscribe(ClockSyncSubscriberItf& subscriber) = 0;
```

File: src/clocksync/clocksync.hpp [hunk 2]
```diff
@@ -53,15 +63,29 @@ class ClockSyncItf {
...
+
+    /**
+     * Unsubscribes from clock sync notifications.
+     *
+     * @param subscriber clock sync notification subscriber.
+     */
+    virtual void Unsubscribes(ClockSyncSubscriberItf& subscriber) = 0;
```
Comment (mlohvynenko) [inline src/clocksync/clocksync.hpp:88]:
```suggestion
    virtual void Unsubscribe(ClockSyncSubscriberItf& subscriber) = 0;
```

File: src/communication/pbhandler.cpp [hunk 3]
```diff
@@ -0,0 +1,160 @@
...
+
+            if (mStop) {
+                break;
+            }
+        }
+
+        auto err = mChannel->Connect();
+        if (!err.IsNone()) {
+            LOG_ERR() << "Failed to connect: name=" << mName << ", err=" << err;
+
```
Comment (mlohvynenko) [inline src/communication/pbhandler.cpp:123]: should we add empty line before continue
if yes, add them in if statments below (in the while) loop
If no, remove this one

File: src/communication/pbhandler.hpp [hunk 4]
```diff
@@ -0,0 +1,94 @@
...
+     * Initializes protobuf handler.
+     *
+     * @param channel communication channel.
+     * @return Error
+     */
+    Error Init(const String& name, ChannelItf& channel);
+
+    /**
+     *  Starts protobuf handler.
+     */
```
Comment (mlohvynenko) [inline src/communication/pbhandler.hpp:37]: * @return Error

File: src/communication/pbhandler.hpp [hunk 5]
```diff
@@ -0,0 +1,94 @@
...
+    Error Start();
+
+    /**
+     * Stops protobuf handler.
+     */
```
Comment (mlohvynenko) [inline src/communication/pbhandler.hpp:42]: * @return Error

File: src/smclient/openhandler.cpp [hunk 6]
```diff
@@ -0,0 +1,116 @@
...
+ **********************************************************************************************************************/
+
+Error OpenHandler::Init(communication::ChannelItf& channel, clocksync::ClockSyncItf& clockSync)
+{
+    auto err = PBHandler::Init("SM open", channel);
+    if (!err.IsNone()) {
+        return AOS_ERROR_WRAP(err);
+    }
+
+    if (!(err = Start()).IsNone()) {
```
Comment (mlohvynenko) [inline src/smclient/openhandler.cpp:28]:
```suggestion
    if (err = Start(); !err.IsNone()) {
```

File: src/smclient/openhandler.cpp [hunk 7]
```diff
@@ -0,0 +1,116 @@
...
+    auto status = pb_decode(&stream, &servicemanager_v4_SMIncomingMessages_msg, &mIncomingMessages);
+    if (!status) {
+        return AOS_ERROR_WRAP(Error(ErrorEnum::eRuntime, "failed to decode message"));
+    }
+
+    Error err;
+
+    switch (mIncomingMessages.which_SMIncomingMessage) {
+    case servicemanager_v4_SMIncomingMessages_clock_sync_tag:
+        if (!(err = ProcessClockSync(mIncomingMessages.SMIncomingMessage.clock_sync)).IsNone()) {
```
Comment (mlohvynenko) [inline src/smclient/openhandler.cpp:82]:
```suggestion
        if (err = ProcessClockSync(mIncomingMessages.SMIncomingMessage.clock_sync); !err.IsNone()) {
```

Comment (mlohvynenko) [reply src/smclient/openhandler.cpp:82]: same suggestion to if statements below

File: src/smclient/smclient.hpp [hunk 8]
```diff
@@ -17,99 +17,103 @@
...
      * @return Error
      */
-    aos::Error SendMonitoringData(const aos::monitoring::NodeMonitoringData& monitoringData) override;
+    Error SendMonitoringData(const monitoring::NodeMonitoringData& monitoringData) override;
 
     /**
      * Subscribes the provided ConnectionSubscriberItf to this object.
      *
      * @param subscriber The ConnectionSubscriberItf that wants to subscribe.
      */
```
Comment (mlohvynenko) [inline src/smclient/smclient.hpp:85]: * @return Error

File: src/communication/pbhandler.cpp [hunk 9]
```diff
@@ -0,0 +1,160 @@
...
+    LOG_DBG() << "Stop PB handler: name=" << mName;
+
+    {
+        LockGuard lock {mMutex};
+
+        if (!mStarted) {
+            return ErrorEnum::eNone;
+        }
+
+        mStop = true;
```
Comment (mykola-kobets-epam) [inline src/communication/pbhandler.cpp:60]: probably  mStarted flag should be updated

**Review by mykola-kobets-epam — APPROVED** [review]

File: src/clocksync/clocksync.hpp (hunk 1 above)
Comment (al1img) [reply src/clocksync/clocksync.hpp:81]: No, we don't use 's at the end of function verbs

File: src/clocksync/clocksync.hpp (hunk 2 above)
Comment (al1img) [reply src/clocksync/clocksync.hpp:88]: the same

File: src/communication/pbhandler.cpp (hunk 3 above)
Comment (al1img) [reply src/communication/pbhandler.cpp:123]: Yes if more than one line of code before

File: src/clocksync/clocksync.hpp (hunk 1 above)
Comment (al1img) [reply src/clocksync/clocksync.hpp:81]: fixed

File: src/clocksync/clocksync.hpp (hunk 2 above)
Comment (al1img) [reply src/clocksync/clocksync.hpp:88]: fixed.

File: src/communication/pbhandler.hpp (hunk 4 above)
Comment (al1img) [reply src/communication/pbhandler.hpp:37]: fixed

**Review by mlohvynenko — APPROVED** [review]

**Review by mlohvynenko — APPROVED** [review]

**Pull request merged**